### PR TITLE
Expose suspend

### PR DESCRIPTION
**Pull request by davyskiba**

### CODE REVIEW

File: app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt [hunk 1]
```diff
@@ -6,9 +6,20 @@ import android.support.v7.app.AppCompatActivity
...
 import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.coroutines.experimental.CoroutineScope
+import kotlinx.coroutines.experimental.Job
+import kotlinx.coroutines.experimental.launch
+import kotlin.coroutines.experimental.CoroutineContext
 
-class MainActivity : AppCompatActivity(), MainView {
+class MainActivity : AppCompatActivity(), MainView, CoroutineScope {
 
+    lateinit var job: Job
```
Comment (shalomhalbert) [inline app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt]: Are all our presenters in Kotlin? To use this approach appropriately, I believe we'll need to implement 
```
lateinit var job: Job
override val coroutineContext: CoroutineContext
        get() = baseCoroutineScope + job
```
in Presenters/ViewModels, not Activitys/Fragments.

Comment (shalomhalbert) [reply app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt]: We'll also need to set the coroutine scope to `Dispatcher.Main` in the Presenters/ViewModels and for every `suspend fun` we'll need to use either
1. `withContext()` inside the functions
2. Have the function's class's declaration define the `CoroutineScope`

Comment (davyskiba) [reply app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt]: We try to write all new code in Kotlin :)
Our current approach is to have Presenters extend CoroutineScope as you suggested.

Unfortunately, to keep our Presenters testable we need to adjust it a bit...
This pull request suggests moving the scope to Activities/Fragments, and expose suspend functions in presenters. This way scope would be closer to lifecycle

File: app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt [hunk 2]
```diff
@@ -2,30 +2,19 @@ package com.panoprogramowanie.testcoroutines.presentation
...
-
-
-    private val baseCoroutineScope by lazy { DependencyInjector.getUIDispatcher() }
     private val presenter by lazy { DependencyInjector.getMainPresenter() }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-        presenter.attach(this)
+        launch { presenter.attach(this@MainActivity) }
```
Comment (shalomhalbert) [inline app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt:17]: @davyskiba the issue I have with this approach is that it breaks MVVM conventions by passing a reference to the Activity to the ViewModel. Is there a way to resolve that?

**Pull request closed without merging**